### ISSUE TITLE
New Check: EmptyLineWrappingInBlockCheck

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#5313]
from #3923 and https://groups.google.com/forum/#!topic/checkstyle/KZn_d8nC4c0
 
New Check
EmptyLineWrappingInBlockCheck
tokens= [CLASS, METHOD, IF, ...........] (by default only class)
topSeparator = [empty-line | empty-line-allowed | no-empty-line]
bottomSeparator = [empty-line | empty-line-allowed | no-empty-line]

empty-line-allowed == mean no validation.

If you consider all my examples above, we could make config for methods to allow user's empty line at the top and completely forbid it at the bottom.
BlockEmptyLineWrappingCheck
tokens= "METHOD_DEF"
topSeparator = "empty-line-allowed"
bottomSeparator = "no-empty-line"

**Comment by ati90ati** [comment]
Hello,

I seen than some changes for AfterLeftCurly and BeforeRightCurly is merged. Could that be used?
Because I did not see it in the documentation.
I want to achieve a check to leave empty space after class def and before last curly of the class.
So this will be valid:

public class MyClass {

// my fields and methods

}

Same for enums and interfaces.

Regards,
Attila

**Comment by ati90ati** [comment]
Thank you.
I'll write here my quick solution, maybe it will help others.

    <module name="RegexpMultiline">
        <property name="format" value="^([^\n ]+ )*(class|interface|enum) [^{]*\{\n[^\n}]"/>
        <property name="message" value="Leave empty row after class/interface/enum definition!"/>
        <property name="severity" value="error"/>
        <property name="fileExtensions" value="groovy,java"/>
    </module>

    <module name="RegexpMultiline">
        <property name="format" value="[^\n{]\n\}\n"/>
        <property name="message" value="Leave empty row before end of class/interface/enum!"/>
        <property name="severity" value="error"/>
        <property name="fileExtensions" value="groovy,java"/>
    </module>

**Comment by Gaurav-Punjabi** [comment]
I'm on it.

**Comment by Gaurav-Punjabi** [comment]
@romani What is the difference between `empty-line` and `empty-line-allowed` option?

**Comment by pbludov** [comment]
> @romani What is the difference between `empty-line` and `empty-line-allowed` option?

Both properties `top/bottom separator` has three options:
* do nothing (allow-empty-line)
* requre (empty-line)
* forbid (no-empty-line)

The default config should process only classes/interfaces/annotations/enums and require that there is one empty line after opening `{` and one empty line before closind `}`.

The following config should enforce that methods/constructors has NO empty line in the method body:
```xml
<module name="EmptyLineWrappingInBlock">
  <property name="tokens" value="CTOR_DEF,METHOD_DEF"/>
  <property name="topSeparator" value="no-empty-line"/>
  <property name="bottomSeparator" value="no-empty-line"/>
</module>
```

Also note that this check should ignore empty blocks at all, since the empty block will be always violated when `topSeparator` is set to `no-empty-line` but `bottomSeparator` is set to its opposite `empty-line` (or vice versa). Also, some coding styles (Google for example) allows them to be inline: `{}`.

**Comment by Gaurav-Punjabi** [comment]
Okay, understood.
Thanks 👍

**Comment by romani** [comment]
there is some progress on fix of this issue, but abandoned - https://github.com/checkstyle/checkstyle/pull/8176

if somebody want to reuse this work to fix issue and send new PR, please be welcome.

**Comment by Rahulkhinchi03** [comment]
Can I work on it @romani sir?

**Comment by Gaoyan1999** [comment]
I'm on this. :)

**Comment by Gaoyan1999** [comment]
Just to confirmation: with empty_line, do we require exactly one empty line—i.e. would multiple consecutive blank lines also count as a violation? Thanks.

**Comment by ati90ati** [comment]
Side note for your last question, there is this check related to that, in this config the check fails for multiple rules (I usually prefer one row only, not more, but it's up to you and to the repo owner to decide what is best for this new check):

```xml
        <module name="EmptyLineSeparator">
            <property name="allowNoEmptyLineBetweenFields" value="true"/>
            <property name="allowMultipleEmptyLines" value="false"/>
            <property name="allowMultipleEmptyLinesInsideClassMembers" value="false"/>
        </module>
```

```
... 'VARIABLE_DEF' has more than 1 empty lines before. [EmptyLineSeparator]
... '}' has more than 1 empty lines after. [EmptyLineSeparator]
```

**Comment by ZaheerAhmadDev** [comment]
I am on it
